### PR TITLE
feat: operator set root

### PR DESCRIPTION
**Pull request by 0x0aa0**
Adds a running hash of stake updates for use in storage proof

### CODE REVIEW

File: src/StakeRegistry.sol [hunk 1]
```diff
@@ -273,6 +273,14 @@ contract StakeRegistry is StakeRegistryStorage {
         }
     }
 
+    /**
+     * @notice clears the quorum root for the given quorum
+     * @dev This function is called by the registry coordinator on stake updates to consolidate the size of the storage proof
+     */
+    function clearQuorumRoot(uint8 quorumNumber) external onlyRegistryCoordinator quorumExists(quorumNumber) {
```
Comment (gpsanant) [inline src/StakeRegistry.sol:280]: emit an event

File: src/StakeRegistry.sol [hunk 2]
```diff
@@ -495,6 +508,12 @@ contract StakeRegistry is StakeRegistryStorage {
         return _totalStakeHistory[quorumNumber].length != 0;
     }
 
+    /// @notice Updates the quorum root for the given quorum 
+    /// @dev quorumRoot+1 = h(quorumRoot, {operator stake update})
+    function _updateQuorumRoot(bytes32 operatorId, uint8 quorumNumber, uint32 blockNumber, uint96 stake) internal {
```
Comment (gpsanant) [inline src/StakeRegistry.sol]: just hash the stake update itself imo, not the individual components

File: src/StakeRegistryStorage.sol [hunk 3]
```diff
@@ -43,6 +43,9 @@ abstract contract StakeRegistryStorage is IStakeRegistry {
      */
     mapping(uint8 => StrategyParams[]) public strategyParams;
 
+    /// @notice mapping from quorum number to the root used for storage proofs of operator set 
+    mapping(uint8 => bytes32) public quorumRoot;
```
Comment (gpsanant) [inline src/StakeRegistryStorage.sol]: quorumOperatorSetRoot

File: src/StakeRegistry.sol (hunk 2 above)
Comment (0x0aa0) [reply src/StakeRegistry.sol]: it will always hash with `nextUpdateBlockNumber: 0` in that case

Comment (gpsanant) [reply src/StakeRegistry.sol]: not when they're deregistering?

File: src/StakeRegistry.sol [hunk 4]
```diff
@@ -347,6 +359,8 @@ contract StakeRegistry is StakeRegistryStorage {
                     nextUpdateBlockNumber: 0,
                     stake: newStake
                 }));
+
+                _updateQuorumRoot(operatorId, quorumNumber, uint32(block.number), newStake);
```
Comment (wadealexc) [inline src/StakeRegistry.sol]: Should be called below, before the function returns. Currently, this misses a branch where the operator's stake is updated twice in one block.

File: src/StakeRegistry.sol (hunk 2 above)
Comment (wadealexc) [reply src/StakeRegistry.sol]: > not when they're deregistering?

also when they're deregistering tho. `nextUpdateBlockNumber` is only set to a nonzero value when we push a new one - so if the hash is over the most recent update, the next update block number will be zero.

File: src/RegistryCoordinator.sol [hunk 5]
```diff
@@ -323,6 +323,7 @@ contract RegistryCoordinator is
                 "RegistryCoordinator.updateOperatorsForQuorum: number of updated operators does not match quorum total"
             );
             address prevOperatorAddress = address(0);
+            stakeRegistry.clearQuorumRoot(quorumNumber);
```
Comment (wadealexc) [inline src/RegistryCoordinator.sol:315]: Are there any cases where we don't want this to be called?

This call here basically means anyone can clear the current quorum root on demand. Is that ok?

File: src/StakeRegistryStorage.sol [hunk 6]
```diff
@@ -43,6 +43,9 @@ abstract contract StakeRegistryStorage is IStakeRegistry {
      */
     mapping(uint8 => StrategyParams[]) public strategyParams;
 
+    /// @notice mapping from quorum number to the root used for storage proofs of operator set 
+    mapping(uint8 => bytes32) public quorumOperatorSetRoot;
```
Comment (8sunyuan) [inline src/StakeRegistryStorage.sol:47]: decrement storage gap for this new mapping slot?

File: src/RegistryCoordinator.sol (hunk 5 above)
Comment (0x0aa0) [reply src/RegistryCoordinator.sol:315]: Yes this ok left permissionless as long as the root is always repopulated with stake updates. This refreshing serves to reduce the size of the corresponding storage proof

File: src/StakeRegistry.sol [hunk 7]
```diff
@@ -359,11 +357,19 @@ contract StakeRegistry is StakeRegistryStorage {
                     nextUpdateBlockNumber: 0,
                     stake: newStake
                 }));
-
-                _updateQuorumRoot(operatorId, quorumNumber, uint32(block.number), newStake);
             }
         }
 
+        _updateQuorumRoot(
```
Comment (gpsanant) [inline src/StakeRegistry.sol:363]: you also need to update where the operator deregisters?

Comment (wadealexc) [reply src/StakeRegistry.sol:363]: This is the method that gets called when they deregister

Comment (wadealexc) [reply src/StakeRegistry.sol:363]: That's why arch was saying this only gets hashed with nextUpdateBlockNumber == 0

File: src/StakeRegistry.sol (hunk 2 above)
Comment (gpsanant) [reply src/StakeRegistry.sol]: oh ok then. no need

File: src/StakeRegistry.sol (hunk 7 above)
Comment (gpsanant) [reply src/StakeRegistry.sol:363]: responded above. sg.

Comment (0x0aa0) [reply src/StakeRegistry.sol:363]: _recordOperatorStakeUpdate() is also called on dereregistration

File: src/StakeRegistry.sol [hunk 8]
```diff
@@ -350,6 +360,16 @@ contract StakeRegistry is StakeRegistryStorage {
...
+            quorumNumber, 
+            operatorId, 
+            StakeUpdate({
+                updateBlockNumber: uint32(block.number),
+                nextUpdateBlockNumber: 0,
```
Comment (gpsanant) [inline src/StakeRegistry.sol:368]: prob don't need this but its ok

**Review by gpsanant — APPROVED** [review]
lgtm

**Comment by wadealexc** [comment]
changed the PR title so the linter is happier when merged

**Comment by teddyknox** [comment]
LGTM! Will verify that storage proofs match `<StakeRegistry contract address>.quorumOperatorSetRoot[quorumNumber]`, and then will verify that the value matches `keccak256(abi.encode(quorumOperatorSetRoot[quorumNumber], operatorId, operatorStakeUpdate))`.

**Review by teddyknox — APPROVED** [review]

**Review by wadealexc — CHANGES_REQUESTED** [review]
Per conversation in Slack, we're going to wait to merge this until it's well tested and documented.

**Pull request closed without merging**